### PR TITLE
[ONNX] Relax sequence tensor dim_param serialization

### PR DESCRIPTION
**Pull request by BowenBao**
Do not assign dim_param for sequence tensor type.
Sequence of tensors could differ in dimension size.
Use a dimension with neither dim_value nor dim_param set
to denote an unknown dimension.
Create and assign dim_param for normal tensor type.

### CODE REVIEW

**Comment by malfet** [comment]
Removing @albanD as reviewer. @BowenBao was it intentional or just a suggested reviewer from github?

**Comment by BowenBao** [comment]
> Removing @albanD as reviewer. @BowenBao was it intentional or just a suggested reviewer from github?

I believe it was auto suggested by github.

**Review by thiagocrepaldi — COMMENTED** [review]
@BowenBao should we have tests with and without tensor sequence and verify dim_param and dim_value are/are not set when applicable?

**Comment by BowenBao** [comment]
> @BowenBao should we have tests with and without tensor sequence and verify dim_param and dim_value are/are not set when applicable?

added unittest

File: torch/csrc/jit/serialization/export.cpp [hunk 1]
```diff
@@ -583,7 +589,7 @@ void GraphEncoder::EncodeValueInfoType(
       // Encode type if either shape or dtype exists.
       onnx::TypeProto_Tensor* onnx_tensor_type =
           onnx_type->mutable_tensor_type();
-      tensorTypeToONNXType(tensor_type, onnx_tensor_type);
+      tensorTypeToONNXType(tensor_type, onnx_tensor_type, tensor_type != n->type());
```
Comment (garymm) [inline torch/csrc/jit/serialization/export.cpp:592]: Please move the comment from lines 563-564 to here and expand it a bit to explain how it relates to `tensor_type != n->type()`. That is, make it clear how `tensor_type != n->type()` relates to sequence types.

File: torch/csrc/jit/serialization/export.cpp [hunk 2]
```diff
@@ -543,7 +543,8 @@ void GraphEncoder::EncodeValueInfoType(
         std::unordered_map<int64_t, std::string>>& dynamic_axes) {
   auto tensorTypeToONNXType = [&dynamic_axes, n, this](
                                   const TensorTypePtr& t,
-                                  onnx::TypeProto_Tensor* onnx_tensor_type) {
+                                  onnx::TypeProto_Tensor* onnx_tensor_type,
+                                  bool is_sequence_tensor) {
```
Comment (garymm) [inline torch/csrc/jit/serialization/export.cpp:547]: I think this would be clearer if we renamed the variable something about what the intended effect is. Something like `bool assign_dim_param`.

File: torch/csrc/jit/serialization/export.cpp [hunk 3]
```diff
@@ -558,7 +559,12 @@ void GraphEncoder::EncodeValueInfoType(
           }
         } else if (sizes[i].is_static()) {
           shape->mutable_dim(i)->set_dim_value(sizes[i].static_size());
-        } else {
+        } else if (!is_sequence_tensor) {
```
Comment (garymm) [inline torch/csrc/jit/serialization/export.cpp:562]: Does the code on line 556-559 need to be guarded by this bool as well?
If not, could you please explain the difference?

File: test/onnx/test_utility_funs.py [hunk 4]
```diff
@@ -602,6 +601,36 @@ def test_error_on_data_parallel(self):
...
+        x = torch.randn(2, 3)
+
+        # Case 1: varying dimension size.
+        f = io.BytesIO()
+        y = torch.randn(2, 4)
+        torch.onnx.export(torch.jit.script(model), (x, y), f, opset_version=self.opset_version,
+                          input_names=['x', 'y'], dynamic_axes={'y': [1]})
+        onnx_model = onnx.load(io.BytesIO(f.getvalue()))
+        loop_output_value_info_proto = onnx_model.graph.output[0]
+        ref_value_info_proto = onnx.helper.make_tensor_sequence_value_info(loop_output_value_info_proto.name, 1, [2, None])
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:622]: Insert a line break somewhere to make this line shorter

File: test/onnx/test_utility_funs.py [hunk 5]
```diff
@@ -602,6 +601,36 @@ def test_error_on_data_parallel(self):
...
+        class Module(torch.nn.Module):
+            def forward(self, x, y):
+                l = [x, y]
+                return l
+
+        model = Module()
+
+        x = torch.randn(2, 3)
+
+        # Case 1: varying dimension size.
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:615]: I think this comment is referring to setting `dynamic_axes` in the export call, if so make it clearer:
```suggestion
        # Case 1: dynamic axis
```

File: test/onnx/test_utility_funs.py [hunk 6]
```diff
@@ -602,6 +601,36 @@ def test_error_on_data_parallel(self):
...
+        f = io.BytesIO()
+        y = torch.randn(2, 4)
+        torch.onnx.export(torch.jit.script(model), (x, y), f, opset_version=self.opset_version,
+                          input_names=['x', 'y'], dynamic_axes={'y': [1]})
+        onnx_model = onnx.load(io.BytesIO(f.getvalue()))
+        loop_output_value_info_proto = onnx_model.graph.output[0]
+        ref_value_info_proto = onnx.helper.make_tensor_sequence_value_info(loop_output_value_info_proto.name, 1, [2, None])
+        self.assertEqual(loop_output_value_info_proto, ref_value_info_proto)
+
+        # Case 2: static dimension size.
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:625]:
```suggestion
        # Case 2: no dynamic axes
```

File: test/onnx/test_utility_funs.py [hunk 7]
```diff
@@ -602,6 +601,36 @@ def test_error_on_data_parallel(self):
                                     "unwrap model from torch.nn.DataParallel. Try "):
             torch.onnx.export(model, x, f, opset_version=self.opset_version)
 
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_loop_sequence_dim(self):
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:605]: What does this have to do with loops?

File: test/onnx/test_utility_funs.py [hunk 8]
```diff
@@ -602,6 +601,36 @@ def test_error_on_data_parallel(self):
...
+                return l
+
+        model = Module()
+
+        x = torch.randn(2, 3)
+
+        # Case 1: varying dimension size.
+        f = io.BytesIO()
+        y = torch.randn(2, 4)
+        torch.onnx.export(torch.jit.script(model), (x, y), f, opset_version=self.opset_version,
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:618]: Why is scripting necessary here?

File: test/onnx/test_utility_funs.py [hunk 9]
```diff
@@ -602,6 +601,36 @@ def test_error_on_data_parallel(self):
...
+                          input_names=['x', 'y'], dynamic_axes={'y': [1]})
+        onnx_model = onnx.load(io.BytesIO(f.getvalue()))
+        loop_output_value_info_proto = onnx_model.graph.output[0]
+        ref_value_info_proto = onnx.helper.make_tensor_sequence_value_info(loop_output_value_info_proto.name, 1, [2, None])
+        self.assertEqual(loop_output_value_info_proto, ref_value_info_proto)
+
+        # Case 2: static dimension size.
+        f = io.BytesIO()
+        y = torch.randn(2, 3)
+        torch.onnx.export(torch.jit.script(model), (x, y), f, opset_version=self.opset_version)
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:628]: ditto on scripting

File: test/onnx/test_utility_funs.py [hunk 10]
```diff
@@ -602,6 +601,36 @@ def test_error_on_data_parallel(self):
                                     "unwrap model from torch.nn.DataParallel. Try "):
             torch.onnx.export(model, x, f, opset_version=self.opset_version)
 
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_loop_sequence_dim(self):
+        class Module(torch.nn.Module):
+            def forward(self, x, y):
+                l = [x, y]
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:608]:
```suggestion
                return [x, y]
```

File: test/onnx/test_utility_funs.py [hunk 11]
```diff
@@ -1067,6 +1096,7 @@ def forward(self, x):
                 return y
 
         x = torch.tensor([1, 2])
+        import caffe2.python.onnx.backend as backend
```
Comment (garymm) [inline test/onnx/test_utility_funs.py:1099]: please add a comment about why this import is here

**Review by garymm — CHANGES_REQUESTED** [review]
@BowenBao You should mention stale PRs like this in our stand-up meetings. This latency is not good!

File: test/onnx/test_utility_funs.py (hunk 7 above)
Comment (BowenBao) [reply test/onnx/test_utility_funs.py:605]: no longer related, removed loop

File: torch/csrc/jit/serialization/export.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/serialization/export.cpp:592]: Changing to type check whether `n` is List type.

File: torch/csrc/jit/serialization/export.cpp (hunk 3 above)
Comment (BowenBao) [reply torch/csrc/jit/serialization/export.cpp:562]: No, because dynamic axis symbol from 556-559 are found directly in `dynamic_axes` argument. This implies this sequence is graph output, and users explicitly assigned that symbol.

File: test/onnx/test_utility_funs.py (hunk 8 above)
Comment (BowenBao) [reply test/onnx/test_utility_funs.py:618]: added comments

        # Export with scripting to keep output as Sequence type.
        # Tracing unpacks the list.

**Comment by BowenBao** [comment]
@garymm Thanks for review. I do want to call out that I did mentioned I had a bunch of items not being reviewed in every stand-ups. I will directly reach out to on-duty individuals next times.

**Comment by BowenBao** [comment]
This PR can be merged into master if #72297 is merged, I will rebase and change the base branch if needed.

File: test/onnx/test_utility_funs.py (hunk 5 above)
Comment (garymm) [reply test/onnx/test_utility_funs.py:616]: Did you forget to make this change?

**Review by garymm — APPROVED** [review]
